src/mds/LogEvent.h

```diff
@@ -49,10 +49,12 @@
 #include "include/buffer_fwd.h"
 #include "include/Context.h"
 #include "include/utime.h"
+#include "include/auto_shared_ptr.h"
 
 class MDSRank;
 class LogSegment;
 class EMetaBlob;
+using AutoSharedLogSegment = auto_shared_ptr<LogSegment>;
```

Comment:
So I'm mystified by the utility of this class. Why couldn't it be:

`using foo = std::shared_ptr<LogSegment>`

?

Reply:
As I mentioned in the commit message, the utility of this class is to automatically convert a pointer to an object that inherits from [`enable_shared_from_this`](https://en.cppreference.com/w/cpp/memory/enable_shared_from_this) into a shared pointer and also adds an implicit conversion from the shared pointer back to a plain pointer. This helps minimize refactoring.

I'll elaborate on how I believe this helps address the original ticket in a separate comment.

Reply:
Normally we use `boost::intrusive_ptr` if other accessors have a raw pointer. That's somewhat inappropriate in a code base we fully control, the mds.

Your autoshared class is interesting though. I'll have to think about this more.